### ISSUE TITLE
File not found after second request

### ISSUE THREAD

**Issue by mbaxter91288** [masbug/flysystem-google-drive-ext#67]
Someone, please tell me what I am doing wrong 🤦 

I have a Laravel app that syncs references to files to the database to create a kind of, file explorer. This has been working fine but now I just can't get things to work. So after literally 3/4 hours of debugging and creating a completely fresh Laravel app with only this package installed (over and above the basics) and I've got this test code:

```
    $files = Storage::disk('google')->files('Renders/Bathroom');
    $file1 = [
        'file' => 'Renders/Bathroom/Bathroom with Black Fixtures.jpg',
        'exists' => Storage::disk('google')->fileExists('Renders/Bathroom/Bathroom with Black Fixtures.jpg')
    ];
    $file2 = [
        'file' => 'Renders/Bathroom/Blue and Grey Bathroom.jpg',
        'exists' => Storage::disk('google')->fileExists('Renders/Bathroom/Blue and Grey Bathroom.jpg')
    ];

    dd([
        'files' => $files,
        'file1' => $file1,
        'file2' => $file2,
    ]);

```

This is the output of the `dd`:

```
^ array:3 [▼
  "files" => array:9 [▼
    0 => "Renders/Bathroom/Bathroom with Black Fixtures.jpg"
    1 => "Renders/Bathroom/Blue and Grey Bathroom.jpg"
    2 => "Renders/Bathroom/Grey and Blue Bathroom.jpg"
    3 => "Renders/Bathroom/Loft Bathroom.jpg"
    4 => "Renders/Bathroom/Marble Bathroom.jpg"
    5 => "Renders/Bathroom/Modern Bathroom.jpg"
    6 => "Renders/Bathroom/Truffle Oak Bathroom.jpg"
    7 => "Renders/Bathroom/White Brick Bathroom.jpg"
    8 => "Renders/Bathroom/White Freestanding Bathroom.jpg"
  ]
  "file1" => array:2 [▼
    "file" => "Renders/Bathroom/Bathroom with Black Fixtures.jpg"
    "exists" => true
  ]
  "file2" => array:2 [▼
    "file" => "Renders/Bathroom/Blue and Grey Bathroom.jpg"
    "exists" => false
  ]
]
```

It seems that as soon as I call `Storage::disk('google')->fileExists('Renders/Bathroom/Bathroom with Black Fixtures.jpg')` once, then any subsequent requests to any other files are not found. To test this further, I simply switched the file paths for `$file1` and `$file2` and I get the same outcome, i.e:

```

    $files = Storage::disk('google')->files('Renders/Bathroom');
    $file1 = [
        'file' => 'Renders/Bathroom/Blue and Grey Bathroom.jpg',
        'exists' => Storage::disk('google')->fileExists('Renders/Bathroom/Blue and Grey Bathroom.jpg')
    ];
    $file2 = [
        'file' => 'Renders/Bathroom/Bathroom with Black Fixtures.jpg',
        'exists' => Storage::disk('google')->fileExists('Renders/Bathroom/Bathroom with Black Fixtures.jpg')
    ];

    dd([
        'files' => $files,
        'file1' => $file1,
        'file2' => $file2,
    ]);

...

^ array:3 [▼
  "files" => array:9 [▼
    0 => "Renders/Bathroom/Bathroom with Black Fixtures.jpg"
    1 => "Renders/Bathroom/Blue and Grey Bathroom.jpg"
    2 => "Renders/Bathroom/Grey and Blue Bathroom.jpg"
    3 => "Renders/Bathroom/Loft Bathroom.jpg"
    4 => "Renders/Bathroom/Marble Bathroom.jpg"
    5 => "Renders/Bathroom/ Modern Bathroom.jpg"
    6 => "Renders/Bathroom/Truffle Oak Bathroom.jpg"
    7 => "Renders/Bathroom/White Brick Bathroom.jpg"
    8 => "Renders/Bathroom/White Freestanding Bathroom.jpg"
  ]
  "file1" => array:2 [▼
    "file" => "Renders/Bathroom/Blue and Grey Bathroom.jpg"
    "exists" => true
  ]
  "file2" => array:2 [▼
    "file" => "Renders/Bathroom/Bathroom with Black Fixtures.jpg"
    "exists" => false
  ]
]
```

This is driving me bananas! Any help would be greatly appreciated!

Thanks,
Mat

**Comment by erikn69** [comment]
Wich laravel version?, package version? php version? give all the info
>This has been working fine but now I just can't get things to work

Did you upgrade the package? still almost no info
if you upgrade, test with the previus version

I did try with the next test and everything is working
```php
/**
 * @test
 */
public function testing_file_exist()
{
    $this->runScenario(function () {
        $adapter = $this->adapter();
        $adapter->write('test/file1.txt', 'contents', new Config());
        $adapter->write('test/file2.txt', 'contents', new Config());
        $adapter->write('test/file3.txt', 'contents', new Config());
        
        $this->assertEquals(true, $adapter->fileExists('test/file1.txt'));
        $this->assertEquals(true, $adapter->fileExists('test/file2.txt'));
        $this->assertEquals(true, $adapter->fileExists('test/file3.txt'));
    });
}
```

**Comment by mbaxter91288** [comment]
@erikn69 thanks for responding so quick! Sorry for lack of information, here's a little more:

PHP Version 8.1.3
"name": "masbug/flysystem-google-drive-ext",
"version": "v2.2.1",

"name": "laravel/framework",
"version": "v9.13.0",

The code is currently working in production. Today I created a new branch locally to upgrade to Nova 4 which I did successfully. I have a custom Tool in Nova that can be triggered by admins which basically syncs the Google drive files so I was reimplementing it from a Tool to an Action.

It was giving me a weird `PDOException: There is no active transaction` which took me down a huge rabbit hole of trial and error debugging. I eventually noticed that when I was getting the file metadata to store in the DB it was throwing this error, I think, because it couldn't find the file in the first place. 

Here is the production code that now no longer seems to work:

```
<?php

namespace App\Actions\Google;

use Carbon\Carbon;
use App\Models\MarketingMaterial;
use Illuminate\Support\Facades\Cache;
use App\Models\MarketingMaterialFolder;
use Illuminate\Support\Facades\Storage;
use Lorisleiva\Actions\Concerns\AsAction;
use App\Actions\Google\GetMarketingFolderNameFromDisplayNameParts;
use App\Actions\Google\GetMarketingFolderParentIdFromDisplayNameParts;

class SyncFilesAndFoldersFromGoogle
{
    use AsAction;

    public function handle(string $folderName = '') : void
    {

        if ($folderName === '') {
            $this->clearDatabaseAndCache();
        }

        foreach (Storage::cloud()->files($folderName) as $fileId) {

            $meta = Storage::cloud()->getAdapter()->getMetadata($fileId);
            $extra = $meta->extraMetadata();

            $parts = explode('/', $extra['display_path']);
    
            $file = MarketingMaterial::create([
                'id' => $extra['id'],
                'name' => $extra['filename'] . '.' . $extra['extension'],
                'filename' => $extra['filename'] . '.' . $extra['extension'],
                'extension' => $extra['extension'],
                'modified_at' => Carbon::parse($meta['lastModified']),
                'mimetype' => $meta['mimeType'],
                'size' => $meta['fileSize'],
                'description' => '',
                'path' => $meta['path'],
                'folder_id' => GetMarketingFolderParentIdFromDisplayNameParts::run($parts),
            ]);

            if (!Storage::disk('thumbnails')->exists($file->getThumbnailFilename())) {
                DownloadFileThumbnailFromGoogle::run( $file );
            }
    
    
        }
        
        foreach (Storage::cloud()->directories($folderName) as $directoryName) {

            // $meta = Storage::cloud()->
            $meta = Storage::cloud()->getAdapter()->getMetadata($directoryName)->extraMetadata();

            // the new Google Drive implementation returns display names so
            // explode it to get it into parts and we'll deal with what we need from there
            $parts = explode('/', $meta['display_path']);

            MarketingMaterialFolder::create([
                'id' => $meta['id'],
                'name' => GetMarketingFolderNameFromDisplayNameParts::run($parts),
                'path' => $meta['display_path'],
                'description' => '',
                'folder_id' => GetMarketingFolderParentIdFromDisplayNameParts::run($parts),
            ]);
    
            SyncFilesAndFoldersFromGoogle::run($directoryName);
    
        }

    }

    /**
     * Clear both the database of MarketingMaterials and MarketingMaterialFolders
     * plus get rid of the cache of marketing view data by tags
     *
     * @return void
     */
    private function clearDatabaseAndCache() : void
    {

        // clear out all existing files and folders
        MarketingMaterial::truncate();
        MarketingMaterialFolder::truncate();

        // clear marketing folder view cache data
        Cache::tags('marketing-folder-data')->flush();

    }

}

```

This code currently works in Production but now locally it will sync the first file, but nothing else. I've tried adding a "if fileExists" check and then if it doesn't, just skip that file and continue - this stops the odd 'no transaction' error but nothing syncs down apart from the first file.

After Googling further, I decided to create a fresh Laravel install today, install the package and create the simple test in the original post which still shows it not finding the file after a subsequent request.

I a fully aware I am probably doing something stupid, but I just can't see it :(

**Comment by erikn69** [comment]
>test with the previus version

**Comment by mbaxter91288** [comment]
@erikn69 ok so I checked the production version and that's `v2.2.0` so I downgraded locally and the file not found issue I am having was resolved, however, I then did get the `Properties can not be manipulated when useHasDir` issue so I manually applied your fixed to the vendor files for now and included a `if (!Storage::disk('google')->fileExists($fileId)) continue;` in my code and it runs successfully. It looks like all the files are synced perfectly!

Is there anything else I can test to help debug?
